tests/Parser/ParserTest.php

```diff
@@ -0,0 +1,125 @@
+<?php
+
+declare(strict_types=1);
+namespace DevCommunityDE\CodeFormatter\Test;
+
+use DevCommunityDE\CodeFormatter\Parser\ElemNode;
+use DevCommunityDE\CodeFormatter\Parser\Node;
+use DevCommunityDE\CodeFormatter\Parser\Parser;
+use DevCommunityDE\CodeFormatter\Parser\TextNode;
+use PHPUnit\Framework\TestCase;
+
+final class ParserTest extends TestCase
+{
+    private function parseFileToArray(string $input)
+    {
+        $parser = new Parser();
+        return iterator_to_array($parser->parseFile($input));
+    }
+
+    private function parseTextToArray(string $input)
+    {
+        $parser = new Parser();
+        return iterator_to_array($parser->parseText($input));
+    }
+
+    public function testParsingAFileBehavesTheSameAsParsingAText()
+    {
+        $input = __DIR__ . '/input-0.txt';
+        $fromText = $this->parseTextToArray(file_get_contents($input));
+        $fromFile = $this->parseFileToArray($input);
+        $this->assertEquals($fromText, $fromFile);
+    }
+
+    public function testPlainBbcodeConsumesOtherBbcodes()
+    {
+        $nodes = $this->parseTextToArray('before plain bbcode[plain][code]test[/code][/plain]');
+        $this->assertCount(2, $nodes);
+        $this->assertInstanceof(TextNode::class, $nodes[0]);
+        $this->assertEquals(Node::KIND_TEXT, $nodes[0]->getKind());
+        $this->assertEquals('before plain bbcode', $nodes[0]->getBody());
+        $this->assertInstanceof(ElemNode::class, $nodes[1]);
+        $this->assertNotTrue($nodes[1]->isCode());
+        $this->assertEquals('[code]test[/code]', $nodes[1]->getBody());
+    }
+
+    public function testBracketsInsideTextAreHandledCorrectly()
+    {
+        $nodes = $this->parseTextToArray('hello[[[[[');
+        $this->assertCount(2, $nodes);
+        $this->assertInstanceof(TextNode::class, $nodes[0]);
+        $this->assertEquals('hello', $nodes[0]->getBody());
+        $this->assertInstanceof(TextNode::class, $nodes[1]);
+        $this->assertEquals('[[[[[', $nodes[1]->getBody());
+
+        $nodes = $this->parseTextToArray('hello[[[[[code]test[/code]');
+        $this->assertCount(3, $nodes);
+        $this->assertInstanceof(TextNode::class, $nodes[0]);
+        $this->assertEquals('hello', $nodes[0]->getBody());
+        $this->assertInstanceof(TextNode::class, $nodes[1]);
+        $this->assertEquals('[[[[', $nodes[1]->getBody());
+        $this->assertInstanceof(ElemNode::class, $nodes[2]);
+        $this->assertTrue($nodes[2]->isCode());
+        $this->assertEquals('test', $nodes[2]->getBody());
+    }
+
+    public function testNestedCodeBbcodesBehaveTheSameAsInXenforo()
+    {
+        $nodes = $this->parseTextToArray('[code][code]test[/code][/code]');
```

Comment:
If possible, I would parse this as an element/code node consisting of a body `[code]test[/code]`.
In the forum, we could then even support syntax highlighting of BB code.

Reply:
I'm afraid this is not possible for the same reason `</script>` cannot occur inside `<script>`. 

For example:
`[code=css].foo::before { content: '[code]test[/code]'; }[/code]`
There is no way to parse this code-element correctly without interpreting the code-content. 
But in order to interpret the code-content, we have to parse the code-element first ... do you see the problem?

Reply:
Ah, see your point. Sure thing.
Thanks for the explanation.

Reply:
It would be possible now after the CODE=rich update (the necessary code to do this is there).
But XenForo doesn't support it ... is there a way to escape [/code] inside a code-element?